File: 04-verbose-hello-world-srv/00-verbose-hello-world-srv-basic/src/main/zava/com/github/jinahya/hello/misc/c02rfc862/Rfc862Tcp3Client.py
#!/usr/bin/python3

###
# #%L
# verbose-hello-world-srv-common
# %%
# Copyright (C) 2018 - 2023 Jinahya, Inc.
# %%
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#      http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# #L%
###

import hashlib
import random
import select
import socket
import sys
from base64 import b64encode
from contextlib import closing

IPv4 = True
# IPv4 = False
FAMILY = socket.AF_INET if IPv4 else socket.AF_INET6
HOST = '127.0.0.1' if IPv4 else "::1"
PORT = 50007


def hook(event, args):
    print(f'\t[audit]: {event} with args={args}')


sys.addaudithook(hook)


# https://stackoverflow.com/a/77426018/330457
def randomize(a):
    a[:] = random.randbytes(len(a))


with closing(socket.socket(FAMILY, socket.SOCK_STREAM)) as s:
    # -------------------------------------------------------------------- bind
    if random.choice([True, False]):
        s.bind((HOST, 0))
        print(f'(optionally) bound to {s.getsockname()}')
    # ----------------------------------------------------------------- connect
    s.connect((HOST, PORT))
    print(f'connected to {s.getpeername()}')
    # --------------------------------------------------------------- configure
    s.setblocking(False)
    # ----------------------------------------------------------------- prepare
    total_bytes_to_send = random.randint(0, 65536)
    print(f'sending {total_bytes_to_send} byte(s)')
    array = bytearray(random.randint(1024, 8192))
    print(f'array.length: {len(array)}')
    hash_ = hashlib.sha256()
    # -------------------------------------------------------------------- loop
    rlist = [s]
    wlist = [s]
    while True:
        r_ready, w_ready, _ = select.select(rlist, wlist, [], 8.0)
        if s in w_ready:
            # ------------------------------------------------------------ send
            randomize(array)
            if total_bytes_to_send < len(array):
                array = array[0:total_bytes_to_send]
            number_of_bytes_written = s.send(array)
            hash_.update(array[0:number_of_bytes_written])
            total_bytes_to_send -= number_of_bytes_written
            if total_bytes_to_send == 0:
                # --------------------------------------------- shutdown-output
                s.shutdown(socket.SHUT_WR)
                wlist = []
        if s in r_ready:
            # ------------------------------------------------------------ recv
            bytes_received = s.recv(len(array))
            if len(bytes_received) == 0:
                break
    # ------------------------------------------------------------ print-digest
    print(f'digest: {b64encode(hash_.digest()).decode()}')

File: 04-verbose-hello-world-srv/00-verbose-hello-world-srv-basic/src/main/java/com/github/jinahya/hello/misc/c02rfc862/Rfc862Tcp3Client.java
package com.github.jinahya.hello.misc.c02rfc862;

/*-
 * #%L
 * verbose-hello-world-srv-common
 * %%
 * Copyright (C) 2018 - 2023 Jinahya, Inc.
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.github.jinahya.hello.util.JavaSecurityMessageDigestUtils;
import com.github.jinahya.hello.util._ExcludeFromCoverage_PrivateConstructor_Obviously;
import lombok.extern.slf4j.Slf4j;

import java.io.EOFException;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.nio.channels.SocketChannel;
import java.util.concurrent.ThreadLocalRandom;

@Slf4j
class Rfc862Tcp3Client extends Rfc862Tcp {

    public static void main(final String... args) throws Exception {
        try (var selector = Selector.open();
             var client = SocketChannel.open()) {
            // -------------------------------------------------------------- configure-non-blocking
            client.configureBlocking(false); // <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
            // ---------------------------------------------------------------------- bind(optional)
            if (ThreadLocalRandom.current().nextBoolean()) {
                logBound(client.bind(new InetSocketAddress(HOST, 0)));
            }
            // --------------------------------------------------------------- connect(try)/register
            final SelectionKey clientKey;
            if (client.connect(ADDR)) {
                clientKey = logConnected(client).register(
                        selector,              // <sel>
                        SelectionKey.OP_WRITE, // <ops>
                        newBuffer().limit(0)   // <att>
                );
            } else {
                clientKey = client.register(
                        selector,               // <sel>
                        SelectionKey.OP_CONNECT // <ops>
                );
            }
            // ----------------------------------------------------------------------------- prepare
            final var digest = newDigest();
            var bytes = logClientBytes(newRandomBytes());
            // ---------------------------------------------------------------------- select/process
            while (clientKey.isValid()) {
                // -------------------------------------------------------------------------- select
                if (selector.select() == 0) {
                    continue;
                }
                final var selectedKeys = selector.selectedKeys();
                final var selectedKey = selectedKeys.iterator().next();
                assert selectedKey == clientKey;
                selectedKeys.clear();
                // ----------------------------------------------------------------- connect(finish)
                if (selectedKey.isConnectable()) {
                    final var channel = (SocketChannel) selectedKey.channel();
                    assert channel == client;
                    if (channel.finishConnect()) {
                        logConnected(channel);
                        selectedKey.interestOpsAnd(~SelectionKey.OP_CONNECT);
                        selectedKey.attach(newBuffer().limit(0));
                        selectedKey.interestOpsOr(SelectionKey.OP_WRITE);
                        assert !selectedKey.isWritable();
                    }
                }
                // --------------------------------------------------------------------------- write
                if (selectedKey.isWritable()) {
                    final var channel = (SocketChannel) selectedKey.channel();
                    assert channel == client;
                    final var buffer = (ByteBuffer) selectedKey.attachment();
                    if (!buffer.hasRemaining()) {
                        ThreadLocalRandom.current().nextBytes(buffer.array());
                        buffer.clear().limit(Math.min(buffer.limit(), bytes));
                    }
                    assert buffer.hasRemaining() || bytes == 0;
                    final var w = channel.write(buffer);
                    assert w >= 0; // why?
                    JavaSecurityMessageDigestUtils.updateDigest(digest, buffer, w);
                    bytes -= w;
                    if (bytes == 0) {
                        // --------------------------------------------------------- shutdown-output
                        channel.shutdownOutput();
                        // --------------------------------------------------------------------- log
                        logDigest(digest);
                    }
                    selectedKey.interestOpsAnd(~SelectionKey.OP_WRITE);
                    buffer.flip();
                    assert buffer.hasRemaining(); // why?
                    selectedKey.interestOpsOr(SelectionKey.OP_READ);
                    assert !selectedKey.isReadable();
                }
                // ---------------------------------------------------------------------------- read
                if (selectedKey.isReadable()) {
                    final var channel = (SocketChannel) selectedKey.channel();
                    assert channel == client;
                    final var buffer = (ByteBuffer) selectedKey.attachment();
                    assert buffer.hasRemaining(); // why?
                    final int r = channel.read(buffer);
                    if (r == -1) {
                        throw new EOFException("unexpected eof");
                    }
                    assert r >= 0; // why?
                    if (!buffer.hasRemaining()) {
                        selectedKey.interestOpsAnd(~SelectionKey.OP_READ);
                        if (bytes == 0) {
                            channel.close();
                            assert !selectedKey.isValid();
                            return;
                        }
                        buffer.limit(Math.min(buffer.limit(), bytes));
                        selectedKey.interestOpsOr(SelectionKey.OP_WRITE);
                        assert !selectedKey.isWritable();
                    }
                }
            }
        }
    }

    @_ExcludeFromCoverage_PrivateConstructor_Obviously
    private Rfc862Tcp3Client() {
        throw new AssertionError("instantiation is not allowed");
    }
}
